test/functional/lua/snippet_spec.lua

```diff
@@ -0,0 +1,336 @@
+          exec_lua('vim.snippet.expand(...)', table.concat({
+            'class ${1:ClassName} {',
+            '\tpublic $1($2) {',
+            '\t\t${3}',
+            '\t}',
+            '}',
+            ''
+          }, '\n'))
+          feed('ModifiedClassName<C-l>')
+          eq({
+            'class ModifiedClassName {',
+            '\tpublic ModifiedClassName() {',
+            '\t\t',
+            '\t}',
+            '}',
+            '',
+          }, helpers.buf_lines(0))
+        end)
+
+        -- We can't manage `complete(...)` in test.
+        -- it('should insert selected choice', function()
+        --   exec_lua('vim.snippet.expand(...)', table.concat({
+        --     'console.${1|log,info,warn,error|}($2);',
+        --   }, '\n'))
+        --   feed('<C-n><C-n><C-n><C-y>')
+        --   eq({
+        --     'console.warn();',
+        --   }, helpers.buf_lines(0))
+        -- end)
+
+        it('should dispose directly modified non-origin tabstop', function()
+          exec_lua('vim.snippet.expand(...)', table.concat({
+            'class ${1:ClassName} {',
+            '\tpublic $1($2) {',
+            '\t\t${3}',
+            '\t}',
+            '}',
+            ''
+          }, '\n'))
+          eq({
+            'class ClassName {',
+            '\tpublic ClassName() {',
+            '\t\t',
+            '\t}',
+            '}',
+            '',
+          }, helpers.buf_lines(0))
+          feed('<Esc><Cmd>call cursor(2, 9)<CR>ciwDirectlyModified<C-l>')
+          eq({
+            'class ClassName {',
+            '\tpublic DirectlyModified() {',
+            '\t\t',
+            '\t}',
+            '}',
+            '',
+          }, helpers.buf_lines(0))
+        end)
+
+        it('should restore the state with undo', function()
+          exec_lua('vim.snippet.expand(...)', table.concat({
+            'class ${1:ClassName} {',
+            '\tpublic $1($2) {',
+            '\t\t${3}',
+            '\t}',
+            '}',
+            ''
+          }, '\n'))
+          feed('ModifiedClassName<C-l><Tab>argument<Esc>')
+          eq({
+            'class ModifiedClassName {',
+            '\tpublic ModifiedClassName(argument) {',
+            '\t\t',
+            '\t}',
+            '}',
+            '',
+          }, helpers.buf_lines(0))
+
+          feed('u<C-l>')
+          eq({
+            'class ClassName {',
+            '\tpublic ClassName() {',
+            '\t\t',
+            '\t}',
+            '}',
+            '',
+          }, helpers.buf_lines(0))
+
+          local to_insert = selection == 'exclusive' and 'i' or 'a'
+          feed(('i<S-Tab><C-g>o<Esc>%sModified<C-l>'):format(to_insert))
+          eq({
+            'class ClassNameModified {',
+            '\tpublic ClassNameModified() {',
+            '\t\t',
+            '\t}',
+            '}',
+            '',
+          }, helpers.buf_lines(0))
+        end)
+
+        it('should dispose snippet if edit outside of range', function()
```

Comment:
"should" is always redundant in test names
```suggestion
        it('disposes snippet if edit outside of range', function()
```